### PR TITLE
es6.spec.arrowFunctions

### PR DESCRIPTION
**Pull request by RReverser**
Added transformer against `new` on arrow expressions.

This is required by spec, but made it optional.

### CODE REVIEW

**Comment by sebmck** [comment]
I actually talked about this in my JSConf talk and why it's a deliberate
decision not to support this. You're sacrificing different spec compliance
that's arguably more serious (the name of a function) in favor of
non-newable arrow functions.

On Friday, 5 June 2015, Ingvar Stepanyan <notifications@github.com> wrote:

> Added transformer against new on arrow expressions.
>
> This is required by spec, but made it optional.
> ------------------------------
> You can view, comment on, or merge this pull request online at:
>
>   https://github.com/babel/babel/pull/1692
> Commit Summary
>
>    - [es6.spec.noNewArrow] Added transformer against `new` on arrow
>    expression.
>
> File Changes
>
>    - *M* src/babel/transformation/file/index.js
>    <https://github.com/babel/babel/pull/1692/files#diff-0> (1)
>    - *A* src/babel/transformation/templates/helper-new-arrow-check.js
>    <https://github.com/babel/babel/pull/1692/files#diff-1> (5)
>    - *M* src/babel/transformation/transformers/es6/arrow-functions.js
>    <https://github.com/babel/babel/pull/1692/files#diff-2> (1)
>    - *A* src/babel/transformation/transformers/es6/spec.no-new-arrow.js
>    <https://github.com/babel/babel/pull/1692/files#diff-3> (19)
>    - *M* src/babel/transformation/transformers/index.js
>    <https://github.com/babel/babel/pull/1692/files#diff-4> (1)
>    - *A*
>    test/core/fixtures/transformation/es6.arrow-functions/no-new/actual.js
>    <https://github.com/babel/babel/pull/1692/files#diff-5> (2)
>    - *A*
>    test/core/fixtures/transformation/es6.arrow-functions/no-new/expected.js
>    <https://github.com/babel/babel/pull/1692/files#diff-6> (14)
>    - *A*
>    test/core/fixtures/transformation/es6.arrow-functions/no-new/options.json
>    <https://github.com/babel/babel/pull/1692/files#diff-7> (3)
>
> Patch Links:
>
>    - https://github.com/babel/babel/pull/1692.patch
>    - https://github.com/babel/babel/pull/1692.diff
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/babel/babel/pull/1692>.
>


-- 
Sebastian McKenzie

**Comment by RReverser** [comment]
> I actually talked about this in my JSConf talk and why it's a deliberate decision not to support this.

I was writing to you in Gitter about that :P

> You're sacrificing different spec compliance that's arguably more serious (the name of a function) in favor of non-newable arrow functions.

It's sacrificed only for anonymous functions. Do you want me to fix it for them, too?

**Comment by RReverser** [comment]
Btw, why do you think that function name (especially absence of it in anonymous function) is more serious than runtime semantic?

**Comment by sebmck** [comment]
Function name is more likely to be relied upon, `new`, not so much. I'm planning on removing all of the optional transformers by either turning them on by default or shifting them into plugins and it's unlikely this would be enabled by default. Would you be up for making it a plugin?

**Comment by RReverser** [comment]
Remove `.prototype` from arrow functions as per spec using `.bind(...)` trick.

**Comment by tejasmanohar** [comment]
+1

File: src/babel/traversal/path/ancestry.js [hunk 1]
```diff
@@ -62,11 +62,12 @@ export function inType(types) {
  * Description
  */
 
-export function inShadow() {
+export function inShadow(key) {
   var path = this;
   while (path) {
     if (path.isFunction()) {
-      if (path.node.shadow) {
+      var {shadow} = path.node;
```
Comment (sebmck) [inline src/babel/traversal/path/ancestry.js]: Spaces around `shadow`.

File: test/core/fixtures/transformation/es6.arrow-functions/spec/options.json [hunk 2]
```diff
@@ -0,0 +1,3 @@
+{
+	"optional": ["es6.spec.arrowFunctions"]
```
Comment (sebmck) [inline test/core/fixtures/transformation/es6.arrow-functions/spec/options.json]: Set `externalHelpers: true` here so the helper isn't present in the expected test.

File: src/babel/transformation/transformers/es6/spec.arrow-functions.js [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+
+export var metadata = {
+  optional: true
+};
+
+export function ArrowFunctionExpression(node, parent, scope, file) {
+  if (node.shadow) return;
+  node.shadow = { this: false };
+
+  var {id} = node;
```
Comment (sebmck) [inline src/babel/transformation/transformers/es6/spec.arrow-functions.js]: Spaces around `id`.

**Comment by RReverser** [comment]
Fixed things & optimized transformer so that it doesn't create variables for arrow functions and doesn't use `instanceof` anymore.

**Comment by RReverser** [comment]
Can I merge this now? It starts to be taking forever :P

**Comment by sebmck** [comment]
I don't know whether or not I want this in core. Still unsure and need time to make a decision.

**Comment by RReverser** [comment]
Oh man :/ This starts feeling like a monopoly.

**Comment by sebmck** [comment]
Why?

**Comment by RReverser** [comment]
Well... because earlier the goal seemed to be to cover the possible maximum of ES6+, even if some parts of it will be optional (like TCO or other things that complicate the code), and now you are abandoning / having doubts about spec-compliant transformations (this one, module prototype) from core while putting own experimental optimization transformers into it at the same time.

Both types of transformers are optional, but one is just extending support for spec features / compliancy, and another one is even somewhat dangerous, and yet the second one wins (makes it's way into the core).

**Comment by RReverser** [comment]
I just wish there would be more transparency on the process of contributions. If you want to move all the optional transformers out of core, as I said, I don't mind, but then it needs to be the same rule for everyone. And if we're keeping them in, then please let's add features that make Babel's support for ES6+ better, too.

**Comment by sebmck** [comment]
> Well... because earlier the goal seemed to be to cover the possible maximum of ES6+, even if some parts of it will be optional (like TCO or other things that complicate the code), and now you are abandoning / having doubts about spec-compliant transformations (this one, module prototype) from core while putting own experimental optimization transformers into it at the same time.

No because Babel is at the point where it might end up not having **any** transformers in the core. Adding new transformers is going completely opposite to that. It shouldn't be a surprise that Babel is moving in this direction, I discussed this extensively back in #568.

> I just wish there would be more transparency on the process of contributions. If you want to move all the optional transformers out of core, as I said, I don't mind, but then it needs to be the same rule for everyone.

It is, a new non-syntax transformer hasn't been added since like 5.0.0 when the React JSX optimisation ones were added. Since then only existing transformers have been renamed or split up and the introduction of new transformers has been restricted to ones in relation to new syntax.

If you're upset about your contribution not being accepted instantly then it should have been discussed before it was made :confused: It puts me in an extremely awkward position where I'm pressured to accept additions that I'm not 100% confident in. I even bought this up in #1347 where I said:

> **Note about pull requests**
>
> Please note that before any significant contributions (feature additions etc) I'd love if an issue was open prior discussing what it is you want to do. There's nothing worse than having to turn down a PR if it could have been avoided prior to all the effort being put in.

Happy to address any specific concerns you have as you seem to be alluding to some criticism of the management structure.

**Comment by RReverser** [comment]
> No because Babel is at the point where it might end up not having any transformers in the core.

Well, now it's weird because I remember when we discussed this and I tried to convince you in splitting the transformation system and transformers themselves into various repos, you were pretty opposed to the idea and wanted to go with "batteries included" approach.

> It shouldn't be a surprise that Babel is moving in this direction, I discussed this extensively back in #568.

I wasn't mentioned / manually subscribed, thus didn't follow the thread, but here is what I see there:

> Yeah, I'm leaning more towards a powerful transformer API with everything else built-in. Meaning that all official transformers would be in the main package by default and only third party ones would be available elsewhere.

or:

> All the current transformers would be included out of the box, I'm not comfortable with splitting everything up

This doesn't sound like "might end up not having any transformers in the core".

Where is the truth? Somewhere in between?

> If you're upset about your contribution not being accepted instantly then it should have been discussed before it was made :confused:

No, definitely not about this, but about priorities of spec-compliant transformers vs experiments currently going directly into 6.0.0 branch.

> it should have been discussed before it was made :confused:

We did, and last time you seemed convinced that keeping such things as plugins is currently much more complicated; also when I proposed to add also support for `.prototype` removal from arrow functions, you said "sure" so I continued to optimize it. Moreover - you were giving code feedbacks, to which I responded by fixing, so it's awkward situation for me as well (like, when someone asks to fix some things on PR, it looks like yet another confirmation of being interested).

**Comment by sebmck** [comment]
> Well, now it's weird because I remember when we discussed this and I tried to convince you in splitting the transformation system and transformers themselves into various repos, you were pretty opposed to the idea and wanted to go with "batteries included" approach.

You were arguing for everything to be split up into modules, scope tracking, traversal etc so other modules could use it. Completely separate thing.

> I wasn't mentioned / manually subscribed, thus didn't follow the thread

You weren't mentioned because you weren't really involved with Babel at the time.

> Where is the truth? Somewhere in between?

Priorities change over time. I was referencing the generic transformation platform direction discussed in that issue.

> We did, and last time you seemed convinced that keeping such things as plugins is currently much more complicated;

Yeah but it seems rather rash to do just do it in core because it's more difficult/impossible to do it in a plugin. If anything this just suggests that plugins need to be made more powerful, not that this should only go in core.

> Moreover - you were giving code feedbacks, to which I responded by fixing, so it's awkward situation for me as well (like, when someone asks to fix some things on PR, it looks like yet another confirmation of being interested).

Yes because I am interested however I'm unsure on how to proceed overall for the project so I'm delaying this PR until I've figured that out. This PR has been open for 5 days, it's seriously not long at all, and I need time to consider the implications. I've been burnt before by accepting PRs that have lead to extreme maintenance hassle (serious bugs etc) and general philosophical fudging (ie. doing something counterintuitive to the present/planned future project goals).

**Comment by RReverser** [comment]
> You were arguing for everything to be split up into modules, scope tracking, traversal etc so other modules could use it. Completely separate thing.

Not completely, just opposite end of your thoughts at that time. Glad we're both moving towards smth in the middle, it gives a hope! :)

> You weren't mentioned because you weren't really involved with Babel at the time.

Of course - it was just respond to "shouldn't be a surprise" - since I wasn't involved at that time yet, I couldn't follow that discussion.

> Yeah but it seems rather rash to do just do it in core because it's more difficult/impossible to do it in a plugin.

Not the only reason - the more important is that it's directly language/spec-related feature.

> Yes because I am interested however I'm unsure on how to proceed overall for the project so I'm delaying this PR until I've figured that out.

To conclude - I definitely don't mind waiting and open to further discussions if it will end up with clear vision of what Babel targets to do, how do we split transformers, how do we ensure that one can easily get language support level that suites his needs and so on and so on.

As well as you, I believe that time could be spent more productively than discussing those questions under each language feature separately, and that we can find some common denominator that would suite everyone's needs.

Cheers! :open_hands:

**Comment by sebmck** [comment]
For sure, also FYI I intend to be more open about development and community feedback/input/help in the #development channel on Slack.

**Comment by sebmck** [comment]
My timeline for 6.0.0 was way too optimistic, it's probably not for a few months. Can't remember what my original objection was but I don't have it now. Merged via ea31229d95369ce74ca98a4ce64d7ddb3393dac3. I merged from your branch @RReverser but it had merge conflicts due to the directory changes which I resolved, your commits should still have been retained, I'm not a git wizard so wasn't able to do anything fancy to retain authorship :disappointed: . Thanks! :smile:

**Comment by RReverser** [comment]
@sebmck No problem, thanks for merging.

**Pull request closed without merging**